Index: mootools-1.2.1-core.js
===================================================================
--- mootools-1.2.1-core.js	(revision 2885)
+++ mootools-1.2.1-core.js	(working copy)
@@ -2487,7 +2487,7 @@
 	id: (/#([\w-]+)/),
 	tag: (/^(\w+|\*)/),
 	quick: (/^(\w+|\*)$/),
-	splitter: (/\s*([+>~\s])\s*([a-zA-Z#.*:\[])/g),
+	splitter: (/\s*([^\s~+>\[\(]+((?:\[[^\]]*\]|\([^\)]*\)))*)\s*([\s~+>])/g),
 	combined: (/\.([\w-]+)|\[(\w+)(?:([!*^$~|]?=)(["']?)([^\4]*?)\4)?\]|:([\w-]+)(?:\(["']?(.*?)?["']?\)|$)/g)
 };
 
@@ -2592,14 +2592,14 @@
 	search: function(self, expression, local){
 		var splitters = [];
 
-		var selectors = expression.trim().replace(Selectors.RegExps.splitter, function(m0, m1, m2){
-			splitters.push(m1);
-			return ':)' + m2;
-		}).split(':)');
+ 		var selectors = (expression.trim() + ' ').replace(Selectors.RegExps.splitter, function(m0, m1, m2, m3){
+ 			splitters.push(m3);
+ 			return m1 + ':)';
+ 		}).split(':)');
 
 		var items, filtered, item;
 
-		for (var i = 0, l = selectors.length; i < l; i++){
+ 		for (var i = 0, l = selectors.length - 1; i < l; i++){
 
 			var selector = selectors[i];
 
